docker-storage-setup.sh

```diff
@@ -839,6 +841,108 @@ get_existing_storage_driver() {
   return 1
 }
 
+docker_root_volume_exists() {
+    lvs $VG/$DOCKER_ROOT_LV_NAME > /dev/null 2>&1 && return 0
+    return 1
+}
+
+# Check if docker root volume is already mounted.
+# This return 0 if volume is already mounted on $DOCKER_ROOT_DIR. Returns 1
+# if volumes needs mounting. On all other error conditions it exits.
+
+docker_root_lv_mounted() {
+  local mounts target
+  if ! mounts=$(findmnt -n --first-only --source /dev/$VG/$DOCKER_ROOT_LV_NAME); then
+    return 1
+  fi
+  [ -z "$mounts" ] && return 1
+  # If root volume is mounted, it should be mounted on $DOCKER_ROOT_DIR
+  # otherwise it is an error.
+  target=`echo $mounts | cut -d " " -f1`
+  [ "$target" == "$DOCKER_ROOT_DIR" ] && return 0
+    echo "Docker root volume /dev/$VG/$DOCKER_ROOT_LV_NAME is mounted on $target instead of ${DOCKER_ROOT_DIR}. Unmount it and try again."
+    exit 1
+}
+
+# mount docker root volume on $DOCKER_ROOT_DIR so that next docker start
+# uses this volume setup by the script.
+mount_docker_root_volume() {
+  if ! mount -t xfs /dev/$VG/$DOCKER_ROOT_LV_NAME $DOCKER_ROOT_DIR > /dev/null;then
+    echo "Failed to mount docker root volume /dev/$VG/$DOCKER_ROOT_LV_NAME on ${DOCKER_ROOT_DIR}."
+    return 1
+  fi
+return 0
+}
+
+# Create a logical volume of size specified by first argument. Name of the
+# volume is specified using second argument.
+create_lv() {
+  local data_size=$1
+  local data_lv_name=$2
+
+  # TODO: Error handling when data_size > available space.
+  if [[ $data_size == *%* ]]; then
+    lvcreate -y -l $data_size -n $data_lv_name $VG || return 1
+  else
+    lvcreate -y -L $data_size -n $data_lv_name $VG || return 1
+  fi
+return 0
+}
+
+setup_docker_root_volume() {
+    if [ -z "$DOCKER_ROOT_VOLUME_SIZE" ]; then
+      echo "Specify a valid value for DOCKER_ROOT_VOLUME_SIZE."
+      exit 1
+    fi
+
+    if ! check_data_size_syntax $DOCKER_ROOT_VOLUME_SIZE; then
+      echo "DOCKER_ROOT_VOLUME_SIZE value $DOCKER_ROOT_VOLUME_SIZE is invalid."
+      exit 1
+    fi
+
+    if ! create_lv $DOCKER_ROOT_VOLUME_SIZE $DOCKER_ROOT_LV_NAME; then
+      echo "Failed to create volume $DOCKER_ROOT_VOLUME_SIZE of size ${DOCKER_ROOT_VOLUME_SIZE}."
+      return 1
+    fi
+
+    if ! mkfs -t xfs /dev/$VG/$DOCKER_ROOT_LV_NAME > /dev/null; then
+      echo "Failed to create filesystem on /dev/$VG/${DOCKER_ROOT_LV_NAME}."
+      return 1
+    fi
+
+    if ! mount_docker_root_volume; then
+      return 1
+    fi
+
+    # setup right selinux label first time fs is created. Mount operation
+    # changes the label of directory to reflect the label on root inode
+    # of mounted fs.
+    if ! restore_selinux_context $DOCKER_ROOT_DIR; then
```

Comment:
You could do a mkdir -Z when creating the directory, But this will not create it with the correct label if the directory is /host/var/lib/docker.  I just submitted some changes to container-selinux that would make the creation of this directory work for certain process types.  We probably can get it to work correctly for docker-storage-setup, if we know which domain it is going to run as.  If it runs as docker_t these directories will get created with the correct label,